### PR TITLE
Add linking to a Global Headquarter for Company.

### PR DESCRIPTION
**Pull request by elcct**
Issue number: DH-1469, DH-1567

### Description of change

This change adds `global_headquarter` to the `Company` model which lets user link Company to its Global Headquarter. It also adds reverse relation `subsidiaries` that allows to list all linked companies.

This need has been discovered through user research, that we need to show a link from any level in the company hierarchy to the global headquarter.

I have also extended `OperatorRule` with `ForeignKeyOperatorRule` that gets value of given field, so that the object linked by foreign key can be passed to the operator, not only the id. I am not sure if the name of this new class is correct, but couldn't come up with anything better.

### Checklist

* [x] Have any relevant search models been updated?
* [x] Have any relevant fixtures (`fixtures/test_data.yaml`) been updated?
* [x] Have any relevant select-/prefetch-related field lists in the views and search apps been updated?
* [x] Has the admin site been updated (for new models, fields etc.)?
* [x] Has the README been updated (if needed)?

### CODE REVIEW

File: datahub/company/models/company.py [hunk 1]
```diff
@@ -130,6 +130,10 @@ class Company(ArchivableModel, BaseModel, CompanyAbstract):
         'self', blank=True, null=True, on_delete=models.SET_NULL,
         related_name='children'
     )
+    global_headquarter = models.ForeignKey(
```
Comment (reupen) [inline datahub/company/models/company.py]: Should be headquarters here (the noun is always headquarters).

File: datahub/company/serializers.py [hunk 2]
```diff
@@ -200,7 +202,10 @@ class CompanySerializer(PermittedFieldsModelSerializer):
         ),
         'uk_establishment_not_in_uk': ugettext_lazy(
             'A UK establishment (branch of non-UK company) must be in the UK.'
-        )
+        ),
+        'global_headquarter_company_is_not_a_global_headquarter': ugettext_lazy(
+            'Company to be set as Global Headquarter must be a Global Headquarter.'
```
Comment (reupen) [inline datahub/company/serializers.py]: The message is a bit confusing :) Maybe it could be a bit more explicit? Also no need for the capital G and H.

**Comment by elcct** [comment]
@reupen 

> Do we need parent/children and global_headquarters/subsidiaries? It's a bit confusing...

Yes, as parent/children would be used to build multi-level hierarchies (if there is going to be a need for that - we are working on this).

File: datahub/company/models/company.py (hunk 1 above)
Comment (elcct) [reply datahub/company/models/company.py]: I should have checked in the dictionary as I wasn't sure about it. Thanks! Renamed now.

File: datahub/company/serializers.py (hunk 2 above)
Comment (elcct) [reply datahub/company/serializers.py]: I have edited the message a bit, but I am not sure how could I write it simpler. This message shouldn't really be shown to the user anyway, because user will be able only to select the link from companies that are global headquarters. Possible edge case is when someone else remove the global headquarters type at the time the other user selects that company to link it.

**Comment by reupen** [comment]
> Yes, as parent/children would be used to build multi-level hierarchies (if there is going to be a need for that - we are working on this).

Yeah, but there's inherent inconsistency in that. Why not get rid of parent until it's actually needed? Then it can be modelled properly.

File: datahub/company/migrations/0025_company_global_headquarter.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# Generated by Django 2.0.2 on 2018-02-12 11:58
```
Comment (marcofucci) [inline datahub/company/migrations/0025_company_global_headquarter.py]: Change name as well?

Comment (elcct) [reply datahub/company/migrations/0025_company_global_headquarter.py]: Nice one! Fixed.

**Comment by reupen** [comment]
> Parent models link to immediate parent company whereas global headquarters models direct link to the company's headquarters. These are two different things. I could deal with parent in another PR.

Isn't the global HQ just the top of the hierarchy? That was my point about consistency. We also now have children and subsidiaries and it's not obvious what the difference is. At the very least the unused stuff could be removed from the API response.

File: datahub/company/serializers.py [hunk 4]
```diff
@@ -354,6 +365,14 @@ class Meta:
                     when=EqualsRule('business_type',
                                     BusinessTypeConstant.uk_establishment.value.id),
                 ),
+                ValidationRule(
+                    'global_headquarters_company_is_not_a_global_headquarters',
```
Comment (reupen) [inline datahub/company/serializers.py]: Should setting a company as its own global HQ be allowed?

Comment (elcct) [reply datahub/company/serializers.py]: It shouldn't be allowed. Added additional validation. Thanks

**Comment by elcct** [comment]
> Isn't the global HQ just the top of the hierarchy?

In many cases it would be, but there are exceptions.

> We also now have children and subsidiaries and it's not obvious what the difference is.

Parent/children is about immediate relationship, maybe this would need either better naming or documentation.

>  At the very least the unused stuff could be removed from the API response.

I have removed parent/children from the API for now.

File: datahub/company/serializers.py [hunk 5]
```diff
@@ -354,6 +363,19 @@ class Meta:
                     when=EqualsRule('business_type',
                                     BusinessTypeConstant.uk_establishment.value.id),
                 ),
+                ValidationRule(
```
Comment (reupen) [inline datahub/company/serializers.py]: Last comment 🙂 

I think both of these could be validators on the fields instead of the class as they don't involve more than one field. That's better as the errors are returned at an earlier stage.

(For the first one, you can get the current instance for update operations via set_context() on your validator.)

**Comment by marcofucci** [comment]
What happens in this scenario:
- A is a GHQ company
- user changes company B with GHQ fk pointing to A
- user change A not to be a GHQ any more

After this change company B still has GHQ pointing to A which is not a GHQ company any longer.

**Comment by elcct** [comment]
@marcofucci we have not decided yet. There are two solutions I have pointed out:

- when user changes a company from GHQ to non GHQ, then all links are being removed.
- when user changes a company from GHQ an error message appears that GHQ has subsidiaries and asks user to remove links manually.
There might be also a need of moving the links from one company to another.

Since we don't know what would be the best solution, we agreed to address this later. In case such event happens before we address it, it could be fixed manually by the user or by a script.

**Comment by marcofucci** [comment]
Hm OK. In the meantime it would be good to notify something when that happens so that we can get an idea of the problem without forgetting it.
It's a bit of a misuse but we could maybe send a warning to sentry so that we get a notification and we know how often it happens?

File: datahub/company/serializers.py [hunk 6]
```diff
@@ -215,7 +224,7 @@ class CompanySerializer(PermittedFieldsModelSerializer):
     business_type = NestedRelatedField(
         meta_models.BusinessType, required=False, allow_null=True
     )
-    children = NestedRelatedField('company.Company', many=True, required=False)
+    subsidiaries = NestedRelatedField('company.Company', many=True, required=False)
```
Comment (reupen) [inline datahub/company/serializers.py]: One more comment actually 😬

I'm not sure about embedding all subsidiaries in the response as there could be a large number in cases. I know limited information is returned here, but then how useful is it?  It seems like something better suited to a separate endpoint with pagination.

Comment (marcofucci) [reply datahub/company/serializers.py]: +1 good point

Comment (elcct) [reply datahub/company/serializers.py]: That makes sense, it should be removed as there will be another endpoint to obtain this data. Fixed.

File: datahub/company/serializers.py (hunk 5 above)
Comment (elcct) [reply datahub/company/serializers.py]: You're right - it looks like doing it that way was a bit of an overkill :) I have replaced it with just the field validation which does the job.

File: datahub/company/serializers.py [hunk 7]
```diff
@@ -264,6 +270,28 @@ class CompanySerializer(PermittedFieldsModelSerializer):
...
+            if (
+                global_headquarters.headquarter_type
+                and str(global_headquarters.headquarter_type.id) != HeadquarterType.ghq.value.id
+            ) or not global_headquarters.headquarter_type:
+                raise serializers.ValidationError(
+                    self.error_messages['global_headquarters_company_is_not_a_global_headquarters']
+                )
+
+            # check if global_headquarters is not pointing to an instance of the model
+            if self.instance and self.instance.id == global_headquarters.id:
```
Comment (reupen) [inline datahub/company/serializers.py]: Could be just `if self.instance == global_headquarters`?

File: datahub/company/serializers.py [hunk 8]
```diff
@@ -264,6 +270,28 @@ class CompanySerializer(PermittedFieldsModelSerializer):
...
+    def validate_global_headquarters(self, global_headquarters):
+        """Ensure that global headquarters is global headquarters and it is not pointing
+        at the model itself.
+        """
+        if global_headquarters:
+            # checks if global_headquarters is global_headquarters
+            if (
+                global_headquarters.headquarter_type
+                and str(global_headquarters.headquarter_type.id) != HeadquarterType.ghq.value.id
+            ) or not global_headquarters.headquarter_type:
```
Comment (reupen) [inline datahub/company/serializers.py]: What about just `global_headquarters.headquarter_type_id != UUID(HeadquarterType.ghq.value.id)` instead of both conditions?

**Comment by reupen** [comment]
> Hm OK. In the meantime it would be good to notify something when that happens so that we can get an idea of the problem without forgetting it.

I agree with @marcofucci here, but I'd say it should just be an error until we decide to do something else. Otherwise we'll end up with objects in an inconsistent state.

**Comment by elcct** [comment]
@reupen @marcofucci so I talked with the team today and we decided to go with an error message when user attempts to change global hq to non global hq when it has subsidiaries. Validation is in the last commit.

File: datahub/company/serializers.py (hunk 7 above)
Comment (elcct) [reply datahub/company/serializers.py]: Good idea. Fixed.

File: datahub/company/serializers.py (hunk 8 above)
Comment (elcct) [reply datahub/company/serializers.py]: Also good idea. Fixed :)

File: datahub/core/validators.py [hunk 9]
```diff
@@ -3,7 +3,6 @@
 from functools import partial
 from operator import eq
 from typing import Any, Callable, Sequence
-
```
Comment (reupen) [inline datahub/core/validators.py]: Why was this removed? :slightly_smiling_face:

File: datahub/company/views.py [hunk 10]
```diff
@@ -48,7 +48,7 @@ class CompanyViewSet(ArchivableViewSetMixin, CoreViewSetV3):
     ).prefetch_related(
         Prefetch('contacts', queryset=get_contact_queryset()),
         Prefetch('investor_investment_projects', queryset=get_slim_investment_project_queryset()),
-        'children',
+        'subsidiaries',
```
Comment (reupen) [inline datahub/company/views.py]: I think we can remove this now since it's not in the response?

File: datahub/core/validators.py (hunk 9 above)
Comment (elcct) [reply datahub/core/validators.py]: That is weird :) Maybe I had focus on wrong window for a moment when deleting. Fixed.

File: datahub/company/views.py (hunk 10 above)
Comment (elcct) [reply datahub/company/views.py]: Of course. Thanks for spotting that! Fixed.
